File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/labs/shared/lab-view.models.ts
import { LabTestComponent } from "./lab-component.model";
import { LabVendorsModel } from "../external-labs/vendors-settings/lab-vendors.model";
import { LabTestRequisition } from "./lab-requisition.model";
import { LabTest } from "./lab-test.model";

//pair of testname and requisitionid.
//needed since we can have multiple tests per sample-code.
//export class TestReqIdPair {
//    public TestName: string = null;
//    public RequisitionId: number = null;
//    public SampleCode: number = null;
//    public OrderDateTime: string = null;
//    //for client side only
//    public isSelected: boolean = true;
//}

export class PatientLabSample {
  public PatientName: string = null;
  public SampleCode: string = null;
  public Specimen: string = null;

  public TestName: string = null;
  public RequisitionId: number = null;
  public OrderDateTime: string = null;
  public OrderStatus: string = null;
  public LastSampleCode: string = null;
  public SampleCreatedOn: string = null;
  public SampleCreatedBy: number = null;
  public LastSpecimenUsed: string = null;

  public RunNumberType: string = "normal";
  public ProviderName: string = null;
  //public SmCode: string = null;
  //public SmNumber: number = null;
  //for client side only
  public IsSelected: boolean = true;
  public UseLastSampleCode: boolean = false;
  public SpecimenList: Array<Object> = null;
  public BarCodeNumber: number = null;
  public HasInsurance: boolean = false;
  public VisitType: string;
  public PatientId: number;
  // public Tests: Array<TestReqIdPair> = null;
  //below is only for client side..
  //public TestCSV: string = null;
  //public SampleCreatedBy: number = null;
  //public SampleCreatedOn: string = null;
}
export class LabTestSpecimen {
  public SpecimenId: number;
  public SpecimenName: string = null;
}

export class PatientTestComponents {
  public LabTestId: number = 0;
  public ComponentJSON: string = null;
  public RequisitionId: number = 0;
  public TestName: string = null;
  public RequestDate: string = null;
  public HasNegativeResults: boolean = false;
  public NegativeResultText: string = null;
}

////
export class LabPendingResultVM {
  public PatientName: string = null;
  public TemplateName: string = null;
  public TemplateId: number = null;
  public PatientId: number = null;
  public Tests: Array<LabTestDetail> = new Array<LabTestDetail>();
  public LabTestCSV: string = null;
  public SampleCode: number = null;
  public SampleDate: string = null;
  public PhoneNumber: string = null;
  public PatientCode: string = null;
  public BarCodeNumber: number = null;
  public WardName: string = null;
}
export class LabTestDetail {
  public RequisitionId: number = 0;
  public TestName: string = null;
  public LabTestId: number = 0;
  public ReportTemplateId: number = 0;
  public ReportTemplateShortName: string = null;
}

//this is like a view model....
export class LabTestViewResult {
  //this is just used to show testname while adding the value of each component
  public TemplateName: string = null;
  public TemplateId: number = null;
  public TestName: string = null;
  public PatientId: number = null;
  public CreatedOn: string = null;
  public PatientName: string = null;
  public SampleCode: number = null;
  public SampleDate: string = null;
  //printId form server-side
  //public PrintId: number = null;
  public LabReportId: number = null;
  public Tests: Array<LabResult_TestVM> = new Array<LabResult_TestVM>();
  //public Components: Array<LabTestComponent> = new Array<LabTestComponent>();
  //public IsPrint: boolean = null;
}

//export class VirtualTestIdNamePair {
//    public TestName: string = null;
//    public RequisitionId: number = 0;
//    public TestId: number = 0;
//    public Components: Array<LabTestComponent> = new Array<LabTestComponent>();
//}

export class LabResult_TemplateVM {
  public TemplateName: string = null;
  public TemplateId: number = null;
  public PatientId: number = null;
  public PatientName: string = null;
  public Tests: Array<LabResult_TestVM> = new Array<LabResult_TestVM>();
}

export class LabResult_TestVM {
  public LabTestId: number = 0;
  public TestName: string = null;
  public ReportingName: string = null;
  public RequisitionId: number = 0;
  public TemplateId: number = 0;
  public TemplateName: string = null;
  public RequestDate: string = null;
  //public PrintId: number = null;
  public LabReportId: number = null;
  public Components: Array<LabTestComponent> = new Array<LabTestComponent>();
  public SelectAll: boolean = false;
  //two properties, one to display the checkbox and second to bind with the checkbox.
  public ShowNegativeCheckbox: boolean = false;
  public IsNegativeResult: boolean = false;
  public NegativeResultText: string = null;
  public ComponentJSON: Array<LabTestComponent> = null;
  public HasNegativeResults: boolean = false;
  public SampleCode: string = null;
  public SampleCreatedOn: string = null;
  public Comments: string = null;
  public TemplateType: string = null; //sud:18Sept (needed to format culture report while mixing culture and normal)
  public ResultFormattedForCulture: Array<any> = new Array<any>(); //sud:18Sept (needed to format culture report while mixing culture and normal)
  public Specimen: string = null;
  public SampleCollectedBy: string = null;
  public SampleCollectedOn: string = null;
  public VendorDetail: LabVendorsModel = null;
  public HasInsurance: boolean = null;
  public BillingStatus: string = "";
  public VerifiedBy: number = null;
  public MaxResultGroup: number = 1;
}

export class LabResult_TemplatesVM {
  public TemplateId: number = null;
  public TemplateName: string = null;
  public HeaderText: string = null;
  public FooterText: string = null;
  public TemplateType: string = null; //sud:18Sept--to loop inside templates
  public TemplateColumns: any = null; //sud:19Sept'18--to group columns in template level.
  public ColCount: number = 0;
  public Tests: Array<LabResult_TestVM> = new Array<LabResult_TestVM>();
}

//Used to add specimen from the addResult page for culture type tests
export class LabTestSpecimenModel {
  public RequisitionId: number = 0;
  public Specimen: string = "";
}


export class LabGroupResultModel { }


export class LabSendSmsModel {
  public BillingStatus: string = null;
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public HasInsurance: boolean = false;
  public IsActive: boolean = true;
  public IsVerified: boolean = false;
  public LOINC: string = null;
  public LabReportId: number = 0;
  public LabTestId: number = 0;
  public LabTestName: string = null;
  public LabTest: Array<LabTest> = [];
  public LabTypeName: string = null;
  public PatientId: number = 0;
  public PatientName: string = null;
  public RequisitionId: number = 0;
  public RunNumberType: string = null;
  public SampleCollectedOnDateTime: string = null;
  public VisitType: string = null;
  public WardName: string = null;
  public SelectAll: boolean = false;
}

File: Code/Components/DanpheEMR.ServerModel/LabModels/LabViewModels.cs
using DanpheEMR.ServerModel.LabModels;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class PatientLabSampleVM
    {
        //pair of testname and requisitionid.
        //needed since we can have multiple tests per sample-code.

        //Changed: Ashim 20Dec2017 
        //Changed as per Requirement Update:  Different tests can have different samplecode
        //SampleCode can be different for different specimen used. 

        //public class TestReqIdPair
        //{
        //    public string TestName { get; set; }
        //    public Int64 RequisitionId { get; set; }

        //}
        public string PatientName { get; set; }
        public string Specimen { get; set; }
        public string TestName { get; set; }
        public Int64 RequisitionId { get; set; }
        public int? SampleCode { get; set; }
        public DateTime? SampleCreatedOn { get; set; }
        public DateTime? OrderDateTime { get; set; }
        public string OrderStatus { get; set; }
        public int? SampleCreatedBy { get; set; }

        public string RunNumberType { get; set; }
        public string LastSpecimenUsed { get; set; }
        public string LastSampleCode { get; set; }
        public string SpecimenList { get; set; }
        public string ProviderName { get; set; }
        public int BarCodeNumber { get; set; }

        public bool? HasInsurance { get; set; }//sud:16Jul'19-- to show insurance flag in collect sample and other pages.

        //public string SmCode { get; set; }
        //public int SmNumber { get; set; }

        //public List<TestReqIdPair> Tests { get; set; }
    }

    public class PatientTestComponentsVM
    {

        public string PatientName { get; set; }
        public string PatientCode { get; set; }
        public int PatientId { get; set; }
        public int ProviderId { get; set; }
        public string ProviderName { get; set; }
        public List<LabTestJSONComponentModel> ComponentJSON { get; set; }
        public int RequisitionId { get; set; }
        public string TestName { get; set; }
        public string LabTestCategory { get; set; }
    }

    public class LabPendingResultVM
    {
        //ashim: 01Sep2018 : Revised since we're now grouping by sample code and labtest array will contain tests of different templates.
        public class LabTestDetail
        {
            public Int64 RequisitionId { get; set; }
            public string TestName { get; set; }
            public Int64 LabTestId { get; set; }
            public int ReportTemplateId { get; set; }
            public string ReportTemplateShortName { get; set; }
            public string RunNumberType { get; set; }
            public int? SampleCollectedBy { get; set; }
            public int? VerifiedBy { get; set; }
            public int? ResultAddedBy { get; set; }
            public int? PrintedBy { get; set; }
            public int? PrintCount { get; set; }
            public string BillingStatus { get; set; }
            public bool ValidTestToPrint { get; set; }
            public int? LabCategoryId { get; set; }
            public int? LabReportId { get; set; }
        }

        public string PatientName { get; set; }
        public string PatientCode { get; set; }
        public DateTime? DateOfBirth { get; set; }
        public string Gender { get; set; }
        //ashim: 01Sep2018 : since we're grouping by sample code, TemplateId and TemplateName is now moved to test level. i.e LabTestDetail
        //public int TemplateId { get; set; }
        //public string TemplateName { get; set; }
        public string PhoneNumber { get; set; }

        public int? SampleCode { get; set; } // used to show the sample code on the pending labresults
        public DateTime? SampleDate { get; set; }
        public string VisitType { get; set; }
        [NotMapped]
        public string SampleCodeFormatted { get; set; }
        public int PatientId { get; set; }
        public string RunNumType { get; set; }
        public bool IsPrinted { get; set; }
        public int? ReportId { get; set; }
        public string BillingStatus { get; set; }
        public int? BarCodeNumber { get; set; }
        public string WardName { get; set; }
        public string ReportGeneratedBy { get; set; }

        public List<LabTestDetail> Tests { get; set; }


        //this variable is used when the final report list data is rendered from mastersearch tab
        public bool IsValidToPrint { get; set; }
        //ashim: 01Sep2018 : since we're grouping by sample code, TemplateId and TemplateName is now moved to test level. i.e LabTestDetail
        //public List<Int64> RequisitionIdList { get; set; }
    }

    /// <summary>
    /// added: sud-22Jun'18
    /// </summary>
    public class LabReportVM
    {
        public string Header { get; set; }
        public ReportLookup Lookups { get; set; }
        public string Columns { get; set; }
        //ashim: 01Sep2018 : since we're grouping by sample code, TemplateId and TemplateName is now moved to test level. i.e LabTestDetail
        //public int TemplateId { get; set; }
        //public string TemplateName { get; set; }
        public int? ReportId { get; set; }
        public string Signatories { get; set; }

        //Added by Anish to show CK Editor based on TemplateType(either 'normal' or 'html')
        public int TemplateId { get; set; }
        public string TemplateType { get; set; }
        public string TemplateHTML { get; set; }
        public string TemplateName { get; set; }
        public string FooterText { get; set; }
        public string Comments { get; set; }
        public bool? IsPrinted { get; set; }
        public bool ValidToPrint { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? ReportCreatedOn { get; set; }
        public int? ReportCreatedBy { get; set; }
        public int? BarCodeNumber { get; set; }
        public DateTime? PrintedOn { get; set; }
        public int? PrintedBy { get; set; }
        public int? PrintCount { get; set; }
        public string PrintedByName { get; set; }
        //make it array of tests or results whatever needed
        public List<LabReportTemplateVM> Templates { get; set; }
    }
    public class ReportLookup
    {
        public int PatientId { get; set; }
        public string PatientName { get; set; }
        public string PatientCode { get; set; }
        public string Gender { get; set; }
        public DateTime? DOB { get; set; }
        public string PhoneNumber { get; set; }
        public string Address { get; set; }
        public int? ReferredById { get; set; }
        public string ReferredBy { get; set; }
        public DateTime? ReceivingDate { get; set; }
        public DateTime? ReportingDate { get; set; }
        public DateTime? SampleDate { get; set; }
        public int? SampleCode { get; set; }
        public string SampleCodeFormatted { get; set; }
        public string VisitType { get; set; }
        public string RunNumberType { get; set; }
    }
    public class LabReportTemplateVM
    {
        public int? TemplateId { get; set; }
        public string TemplateName { get; set; }
        public string TemplateType { get; set; }
        public string TemplateHtml { get; set; }
        public string HeaderText { get; set; }
        public string FooterText { get; set; }
        public string TemplateColumns { get; set; }//sud:19Sept'18--to group columns in template level.
        public object Tests { get; set; }
        public int? DisplaySequence { get; set; }
    }


    //start: sud: 12Sept'18 -- for Lab-Result and Reports formatting
    public class LabResult_Denormalized_VM
    {
        //lookup
        public int PatientId { get; set; }
        public string PatientName { get; set; }
        public string PatientCode { get; set; }
        public string Gender { get; set; }
        public DateTime? DOB { get; set; }
        public string PhoneNumber { get; set; }
        public string Address { get; set; }
        public int? ReferredById { get; set; }
        public string ReferredBy { get; set; }
        public DateTime? ReceivingDate { get; set; }
        public DateTime? ReportingDate { get; set; }
        public DateTime? SampleDate { get; set; }
        public int? SampleCode { get; set; }
        public string SampleCodeFormatted { get; set; }
        public string RunNumberType { get; set; }

        //Lab Report
        public int? LabReportId { get; set; }
        public int? TemplateId { get; set; }
        public bool? IsPrinted { get; set; }
        public string Signatories { get; set; }
        public DateTime? CreatedOn { get; set; }
        public int? CreatedBy { get; set; }
        public DateTime? ReportCreatedOn { get; set; }
        public int? ReportCreatedBy { get; set; }
        public bool? IsActive_Test { get; set; }
        public int? ModifiedBy { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public string ReferredByDr { get; set; }
        public string LabReportComments { get; set; }
        public DateTime? PrintedOn { get; set; }
        public int? PrintedBy { get; set; }
        public int? PrintCount { get; set; }
        public string PrintedByName { get; set; }

        //Requisition + LabTests
        public int? PatientVisitId { get; set; }
        public int? ProviderId { get; set; }
        public Int64 LabTestId { get; set; }
        public string LabTestName { get; set; }
        public string LabTestSpecimen { get; set; }
        public string LabTestSpecimenSource { get; set; }
        public string Urgency { get; set; }
        public DateTime? OrderDateTime { get; set; }
        public string ProviderName { get; set; }
        public string BillingStatus { get; set; }
        public string Specimen { get; set; }
        public string OrderStatus { get; set; }
        public string RequisitionRemarks { get; set; }
        public DateTime? SampleCreatedOn { get; set; }
        public int? SampleCreatedBy { get; set; }
        public string Comments { get; set; }
        public int ReportTemplateId { get; set; }
        public int? DiagnosisId { get; set; }
        public object LabTestComponentsJSON { get; set; }
        public string Description { get; set; }
        public int? TestDisplaySequence { get; set; }
        public bool? HasNegativeResults { get; set; }
        public string NegativeResultText { get; set; }
        public string ReportingName { get; set; }
        public string Interpretation { get; set; }
        public int? VerifiedBy { get; set; }

        //Test Component Result
        public Int64? TestComponentResultId { get; set; }
        public Int64 RequisitionId { get; set; }
        public int ResultingVendorId { get; set; }
        public string Value { get; set; }
        public string Unit { get; set; }
        public string Range { get; set; }
        public string ComponentName { get; set; }
        public string Method { get; set; }
        public string Remarks { get; set; }
        public string RangeDescription { get; set; }
        public bool? IsNegativeResult { get; set; }
        public bool? IsAbnormal { get; set; }
        public bool? IsActive_Component { get; set; }
        public int? ResultGroup { get; set; }


        //LabReportTemplateModel
        public string ReportTemplateShortName { get; set; }
        public string ReportTemplateName { get; set; }
        public bool? IsDefault { get; set; }
        public string HeaderText { get; set; }
        public string ColSettingsJSON { get; set; }
        public string TemplateType { get; set; }
        public string TemplateHTML { get; set; }
        public string FooterText { get; set; }
        public int? TemplateDisplaySequence { get; set; }

        public string VisitType { get; set; }

        public bool? HasInsurance { get; set; }
        public string AbnormalType { get; set; }

    }

    public class LabTest_Temp_VM
    {
        public string TestName { get; set; } //= test.ReportingName,
        public string ReportingName { get; set; }
        public Int64 RequisitionId { get; set; } //= req.RequisitionId,
        public Int64 LabTestId { get; set; }
        public object ComponentJSON { get; set; }
        public bool? HasNegativeResults { get; set; }
        public string NegativeResultText { get; set; }
        //SampleCode = labReq.SampleCode,
        public DateTime? RequestDate { get; set; }
        //SampleCreatedOn = labReq.SampleCreatedOn,
        public string Comments { get; set; }
        public int? DisplaySequence { get; set; }
        public string Interpretation { get; set; }
        public List<LabTestComponentResult> Components { get; set; }

        public string Specimen { get; set; }
        public string SampleCollectedBy { get; set; }
        public DateTime? SampleCollectedOn { get; set; }

        public LabVendorsModel VendorDetail { get; set; }
        public bool? HasInsurance { get; set; }
        public string BillingStatus{ get; set; }
        public int? VerifiedBy { get; set; }
        public int ResultingVendorId { get; set; }
        public int? MaxResultGroup { get; set; }
    }

    //end: sud: 12Sept'18 -- for Lab-Result and Reports formatting

}
